### PR TITLE
CFE-1915: Allow dots in variable identifiers with no such bundle

### PR DESCRIPTION
**Pull request by vpodzime**
As described and discussed in CFE-1915, defining remote variables
(injecting variables into remote bundles) is dangerous and must
be blocked. However, using a dot-separated common prefix for
variables raises no security concerns and can be considered
valid.

Changelog: Commit

### CODE REVIEW

**Comment by vpodzime** [comment]
This needs some extra tests for the now-allowed use case.

**Comment by vpodzime** [comment]
The goal in the changes here was to minimize performance penalty when evaluating policies that don't use this corner-case variable definition use case.

File: libpromises/expand.c [hunk 1]
```diff
@@ -647,6 +647,46 @@ void BundleResolve(EvalContext *ctx, const Bundle *bundle)
         "Resolving classes and variables in 'bundle %s %s'",
         bundle->type, bundle->name);
 
+    /* first check if some variables were injected remotely into this bundle and
+     * remove them (CFE-1915) */
+    const Seq *remote_var_promises = EvalContextGetRemoteVarPromises(ctx);
```
Comment (vpodzime) [inline libpromises/expand.c]: This needs to be limited to only promises that might be injecting variables into this bundle not all the other bundles.

File: libpromises/verify_vars.c [hunk 2]
```diff
@@ -53,18 +54,50 @@ static ConvergeVariableOptions CollectConvergeVariableOptions(EvalContext *ctx,
...
+            Log(LOG_LEVEL_INFO, "Potential remote bundle variable injection detected!");
+            if (StringSetContains(EvalContextGetBundleNames(ctx), prefix))
+            {
+                Log(LOG_LEVEL_ERR, "Remote bundle variable injection detected!");
+                free(prefix);
+                return false;
+            }
+            /* the conflicting bundle may be defined later, we need to remember
+             * this promise as potentially dangerous */
+            EvalContextPushRemoteVarPromise(ctx, pp->org_pp);
```
Comment (vpodzime) [inline libpromises/verify_vars.c]: Do not lose information about the bundle name (``prefix``) here.

File: libpromises/expand.c (hunk 1 above)
Comment (nickanderson) [reply libpromises/expand.c]: Off the top of my head: commands (via modules), classes type promises (some define classic arrays as a side effect), vars, meta, defaults

**Comment by vpodzime** [comment]
[Build Status](https://ci.cfengine.com/job/pr-pipeline/581/)

**Comment by vpodzime** [comment]
Some more tests are probably still needed.

**Comment by vpodzime** [comment]
[Build Status](https://ci.cfengine.com/job/pr-pipeline/592/)

File: libpromises/eval_context.c [hunk 3]
```diff
@@ -3026,3 +3064,48 @@ void EvalContextAllClassesLoggingEnable(EvalContext *ctx, bool enable)
     assert (ctx);
     Nova_ClassHistoryEnable(&(ctx->all_classes), enable);
 }
+
+void EvalContextPushBundleName(const EvalContext *ctx, const char *bundle_name)
```
[the remark this answers is not in the document]
Comment (olehermanse) [reply libpromises/eval_context.c]: Don't pay too much attention to this. It's good.

File: libpromises/expand.c [hunk 4]
```diff
@@ -641,12 +641,73 @@ void BundleResolvePromiseType(EvalContext *ctx, const Bundle *bundle, const char
...
 void BundleResolve(EvalContext *ctx, const Bundle *bundle)
 {
     Log(LOG_LEVEL_DEBUG,
         "Resolving classes and variables in 'bundle %s %s'",
         bundle->type, bundle->name);
 
+    /* first check if some variables were injected remotely into this bundle and
+     * remove them (CFE-1915) */
+    const Seq *remote_var_promises = EvalContextGetRemoteVarPromises(ctx, bundle->name);
+    if ((remote_var_promises != NULL) && SeqLength(remote_var_promises) > 0)
```
Comment (olehermanse) [inline libpromises/expand.c]: Could be nice to put all of this remote var related logic in a separate function.

File: libpromises/expand.c [hunk 5]
```diff
@@ -641,12 +641,73 @@ void BundleResolvePromiseType(EvalContext *ctx, const Bundle *bundle, const char
...
+        /* TODO: this is expensive and should be removed! */
+        for (size_t i = 0; i < promises_length; i++)
+        {
+            const Promise *pp = (Promise *) SeqAt(remote_var_promises, i);
+            VariableTableIterator *iter = EvalContextVariableTableIteratorNew(ctx, NULL, bundle->name, NULL);
+            const Variable *var = VariableTableIteratorNext(iter);
+            while (var != NULL)
+            {
+                /* variables are stored together with their original promises (org_pp) */
+                if (var->promise && var->promise->org_pp == pp)
```
Comment (olehermanse) [inline libpromises/expand.c]: `var->promise != NULL`

File: libpromises/eval_context.c [hunk 6]
```diff
@@ -83,6 +83,27 @@ TYPED_MAP_DEFINE(FuncCache, Rlist *, Rval *,
...
+
+static void SeqDestroy_untyped(void *p)
+{
+    Seq *s = p;
+    SeqDestroy(s);
+}
+
+TYPED_MAP_DECLARE(RemoteVarPromises, char *, Seq *)
+
+TYPED_MAP_DEFINE(RemoteVarPromises, char *, Seq *,
```
[the remark this answers is not in the document]
Comment (vpodzime) [reply libpromises/eval_context.c]: Yes, I was thinking the same. We could do this properly: https://tracker.mender.io/browse/CFE-2828

File: libpromises/eval_context.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (vpodzime) [reply libpromises/eval_context.c]: ``const`` is misleading in C :wink:

**Comment by vpodzime** [comment]
@olehermanse all comments addressed, I believe. And I also added a unittest for the new ``SeqAppendOnce()`` function.

File: libpromises/eval_context.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (olehermanse) [reply libpromises/eval_context.c]: > `const` is misleading in C ;)

Amen.

**Comment by olehermanse** [comment]
@vpodzime Looks like you have a few segfaults in Travis to take care of.

**Comment by vpodzime** [comment]
> @vpodzime Looks like you have a few segfaults in Travis to take care of.

Yeah, turns out map destructors are not good at accepting ``NULL``.

**Comment by vpodzime** [comment]
[Build Status](https://ci.cfengine.com/job/pr-pipeline/602/)

**Comment by vpodzime** [comment]
@olehermanse should be fine again, now.

**Comment by amousset** [comment]
I'm building and trying this in Rudder.

**Comment by amousset** [comment]
Apart from the small Log comment, this works great, thanks!

File: libpromises/eval_context.c [hunk 7]
```diff
@@ -3043,3 +3081,48 @@ void EvalContextAllClassesLoggingEnable(EvalContext *ctx, bool enable)
...
+    if (promises == NULL)
+    {
+        /* initialize the sequence if needed */
+        /* ItemDestroy == NULL because we need to store the exact pointers not
+         * copies */
+        promises = SeqNew(10, NULL);
+        RemoteVarPromisesMapInsert(ctx->remote_var_promises, xstrdup(bundle_name), promises);
+    }
+    /* intentionally not making a copy here, we need the exact pointer */
+    SeqAppend(promises, (void *) pp);
```
Comment (olehermanse) [inline libpromises/eval_context.c]: Why take a const pointer and discard the const?

File: libpromises/verify_vars.c [hunk 8]
```diff
@@ -53,18 +54,50 @@ static ConvergeVariableOptions CollectConvergeVariableOptions(EvalContext *ctx,
...
+    /* we only care about variable name prefix, not dots inside array keys */
+    if ((dot != NULL) && ((bracket == NULL) || (dot < bracket)))
+    {
+        /* detect and prevent remote bundle variable injection (CFE-1915) */
+        char *prefix = xstrndup(var_name, dot - var_name);
+        const Bundle *cur_bundle = PromiseGetBundle(pp);
+
+        if (!StringSafeEqual(prefix, cur_bundle->name))
+        {
+            Log(LOG_LEVEL_INFO, "Potential remote bundle variable injection detected!");
```
[the remark this answers is not in the document]
Comment (vpodzime) [reply libpromises/verify_vars.c]: To be honest, this was just a debugging message I ended up considering useful to keep. I agree it should be just ``VERBOSE``. And it should contain the variable name. Fixing.

File: libpromises/eval_context.c (hunk 7 above)
Comment (vpodzime) [reply libpromises/eval_context.c]: Because we don't have a *const Sequence*. Taking ``const Promise *`` is suggesting that it won't be modified and/or deallocated. But we don't have a sequence type that would provide such guarantees and expressed them in its type signatures. So we have to rely on the implementation when ``ItemDestroy == NULL``. And we want to avoid compiler warnings.

**Comment by vpodzime** [comment]
[Build Status](https://ci.cfengine.com/job/testing-pr/574/)
